glean_parser/csharp.py

```diff
@@ -0,0 +1,166 @@
+                    if not first:
+                        yield ", "
+                    yield from self.iterencode(key)
+                    yield " to "
+                    yield from self.iterencode(subvalue)
+                    first = False
+                yield ")"
+            elif isinstance(value, enum.Enum):
+                yield (value.__class__.__name__ + "." + util.Camelize(value.name))
+            elif isinstance(value, set):
+                yield "setOf("
+                first = True
+                for subvalue in sorted(list(value)):
+                    if not first:
+                        yield ", "
+                    yield from self.iterencode(subvalue)
+                    first = False
+                yield ")"
+            else:
+                yield from super().iterencode(value)
+
+    return "".join(CSharpEncoder().iterencode(value))
+
+
+def type_name(obj: Union[metrics.Metric, pings.Ping]) -> str:
+    """
+    Returns the C# type to use for a given metric or ping object.
+    """
+    generate_enums = getattr(obj, "_generate_enums", [])
+    if len(generate_enums):
+        template_args = []
+        for member, suffix in generate_enums:
+            if len(getattr(obj, member)):
+                template_args.append(util.camelize(obj.name) + suffix)
+            else:
+                if suffix == "Keys":
+                    template_args.append("NoExtraKeys")
+                else:
+                    template_args.append("No" + suffix)
+
+        return "{}<{}>".format(class_name(obj.type), ", ".join(template_args))
+
+    return class_name(obj.type)
+
+
+def class_name(obj_type: str) -> str:
+    """
+    Returns the C# class name for a given metric or ping type.
+    """
+    if obj_type == "ping":
+        return "PingType"
+    if obj_type.startswith("labeled_"):
+        obj_type = obj_type[8:]
+    return util.Camelize(obj_type) + "MetricType"
+
+
+def output_csharp(
+    objs: metrics.ObjectTree, output_dir: Path, options: Dict[str, Any] = {}
+) -> None:
+    """
+    Given a tree of objects, output C# code to `output_dir`.
+
+    :param objects: A tree of objects (metrics and pings) as returned from
+        `parser.parse_objects`.
+    :param output_dir: Path to an output directory to write to.
+    :param options: options dictionary, with the following optional keys:
+
+        - `namespace`: The package namespace to declare at the top of the
+          generated files. Defaults to `GleanMetrics`.
+        - `glean_namespace`: The package namespace of the glean library itself.
+          This is where glean objects will be imported from in the generated
+          code.
+    """
+    template = util.get_jinja2_template(
+        "csharp.jinja2",
+        filters=(
+            ("csharp", csharp_datatypes_filter),
+            ("type_name", type_name),
+            ("class_name", class_name),
+        ),
+    )
+
+    # The object parameters to pass to constructors
+    extra_args = [
```

Comment:
I'm not a fan of this duplication, but we should tackle that globally: https://bugzilla.mozilla.org/show_bug.cgi?id=1652471

Reply:
Oh, thanks for doing this!